File: java/src/civitas/Jugador.java
package civitas;

import java.util.ArrayList;

public class Jugador implements Comparable<Jugador> {
    private final static int CasasMax = 4;
    private final static int CasasPorHotel = 4;
    private boolean encarcelado;
    private static final int HotelesMax = 4;
    private String nombre;
    private int numCasillaActual;
    private final static int pasoPorSalida = 1000;
    private final static int precioLibertad = 200;
    private boolean puedeComprar;
    private float saldo;
    private static float saldoInicial = 7500;
    private Sorpresa salvoconducto;
    private ArrayList<TituloPropiedad> propiedades;

    Jugador(String nombre) {
        this.nombre = nombre;
        encarcelado=false;
        numCasillaActual=0;
        puedeComprar=true;
        saldo=saldoInicial;
        salvoconducto=null;
        propiedades=new ArrayList<TituloPropiedad>();
    }

    protected Jugador(Jugador otro){
        otro.encarcelado=encarcelado;
        otro.nombre=nombre;
        otro.numCasillaActual=numCasillaActual;
        otro.puedeComprar=puedeComprar;
        otro.saldo=saldo;
        otro.salvoconducto=salvoconducto;
        otro.propiedades=propiedades;
    }

    protected boolean debeSerEncarcelado() {
        if (encarcelado) {
            return false;
        }

        else if (tieneSalvoconducto()) {
            perderSalvoconducto();
            Diario.getInstance().ocurreEvento("El jugador " + nombre + " ha perdido su salvoconducto");
            return false;
        }

        return true;
    }

    boolean encarcelar(int numCasillaCarcel) {
        if (debeSerEncarcelado()) {
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.getInstance().ocurreEvento("El jugador " + nombre + " ha sido encarcelado");
        }

        return encarcelado;
    }

    boolean obtenerSalvoconducto(Sorpresa s) {
        if (encarcelado) {
            return false;
        }

        else {
            salvoconducto = s;
            return true;
        }
    }

    private void perderSalvoconducto() {
        salvoconducto.usada();
        salvoconducto=null;
    }

    boolean enBancarrota() {
        return saldo <= 0;
    }

    boolean tieneSalvoconducto() {
        return salvoconducto != null;
    }

    boolean puedeComprarCasilla() {
        puedeComprar = !encarcelado;
        return puedeComprar;
    }

    boolean paga(float cantidad) {
        return modificarSaldo(cantidad * -1);
    }

    boolean pagaImpuesto(float cantidad) {
        if (encarcelado) return false;
        return paga(cantidad);
    }
    
    boolean pagaAlquiler(float cantidad) {
        if (encarcelado) return false;
        return paga(cantidad);
    }
    
    boolean recibe(float cantidad) {
        if (encarcelado) return false;
        return modificarSaldo(cantidad);
        }

    boolean modificarSaldo(float cantidad) {
        saldo += cantidad;
        Diario.getInstance().ocurreEvento("Se ha modificado el saldo en: " + cantidad + ", SALDO ACTUAL: " + saldo);
        return true;
    }

    boolean moverACasilla(int numCasilla) {
        if (encarcelado) {
            return false;
        }
        else {
            numCasillaActual = numCasilla;
            puedeComprar = false;
            Diario.getInstance().ocurreEvento("Se ha desplazado a la casilla: " + numCasilla);
            return true;
        }
    }

    private boolean puedoGastar(float precio) {
        return saldo >= precio;
    }

    boolean vender(int ip) {

        if (!encarcelado && existeLaPropiedad(ip) && propiedades.get(ip).vender(this)) {
                propiedades.remove(ip);
                Diario.getInstance().ocurreEvento("Se ha vendido la propiedad: " + ip);
                return true;
            }

            return false;
    }

    boolean tieneAlgoQueGestionar()  {
        return propiedades.size() != 0;
    }

    private boolean puedeSalirCarcelPagando() {
        return saldo >= precioLibertad;
    }

    boolean salirCarcelPagando() {
        if (encarcelado && puedeSalirCarcelPagando()) {
            paga(precioLibertad);
            encarcelado = false;
            Diario.getInstance().ocurreEvento("Jugador: " + nombre + " ha salido de la carcel PAGANDO");
            return true;
        }
            return false;
    }
    
    boolean salirCarcelTirando() {
        if (Dado.getInstance().salgoDeLaCarcel()) {
            encarcelado = false;
            Diario.getInstance().ocurreEvento("Jugador: " + nombre + " ha salido de la carcel PAGANDO");
            return true;
        }
        else {
            return false;
        }
    }
    
    boolean pasaPorSalida() {
        modificarSaldo(pasoPorSalida);
        Diario.getInstance().ocurreEvento("Jugador: " + nombre + " ha pasado por la salida");
        return true;
    }

    public int compareTo(Jugador otro) {
        if (saldo > otro.saldo) {
            return 1;
        }

        else if (saldo < otro.saldo) {
            return -1;
        }

        else return 0;
    }

    public String toString() {return "";}

    protected String getNombre(){return nombre;}

    public int getNumCasillaActual() {
        return numCasillaActual;
    }
    public boolean isEncarcelado(){
        return encarcelado;
    }

    public boolean getPuedeComprar() {return puedeComprar;}

	public boolean tieneSalvoConducto() {
		return false;
	}

	public void obtenerSalvoConducto() {
	}

	public int cantidadCasasHoteles() {
        int cant=0;
		for (int i=0; i<propiedades.size(); i++){
            cant+=propiedades.get(i).cantidadCasasHoteles();
        }
        return cant;
	}

    boolean cancelarHipoteca(int ip){
        boolean result=false;
        if (encarcelado) return result;
        if (existeLaPropiedad(ip)){
            TituloPropiedad propiedad=propiedades.get(ip);
            float cantidad=propiedad.getImporteCancelarHipoteca();
            boolean puedoGastar=puedoGastar(cantidad);
            if(puedoGastar){
                result=propiedad.cancelarHipoteca(this);
                if(result) Diario.getInstance().ocurreEvento("El jugador "+nombre+" cancela hipoteca de la propiedad "+ip);
            }
        }
        return result;
    }

    boolean comprar(TituloPropiedad titulo){
        boolean result=false;
        if(encarcelado) return result;
        if(puedeComprar){
            float precio=titulo.getPrecioCompra();
            if(puedoGastar(precio)){
                result=titulo.comprar(this);
                if(result)
                propiedades.add(titulo);
                Diario.getInstance().ocurreEvento("El jugador "+nombre+" compra la propiedad "+titulo.toString());
            }
            puedeComprar=false;
        }
        return result;
    }

    boolean construirHotel(int ip){
        boolean result=false;
        if(encarcelado) return result;
        if (!encarcelado){
            TituloPropiedad propiedad=propiedades.get(ip);
            boolean puedoEdificarHotel=puedoEdificarHotel(propiedad);

            if(puedoEdificarHotel){
                result=propiedad.construirHotel(this);
                int casasPorHotel=getCasasPorHotel();
                propiedad.destruirCasas(casasPorHotel, this);
            }
            Diario.getInstance().ocurreEvento("El jugador "+nombre+" ha construido un hotel en "+ip);
        }
        return result;
    }

    boolean construirCasa(int ip){
        boolean result=false;
        if(encarcelado) return result;
        if (!encarcelado){
            TituloPropiedad propiedad=propiedades.get(ip);
            boolean puedoEdificarCasa=puedoEdificarCasa(propiedad);

            if(puedoEdificarCasa){
                result=propiedad.construirCasa(this);
            }
            Diario.getInstance().ocurreEvento("El jugador "+nombre+" ha construido una casa en "+ip);
        }
        return result;
    }
    
    boolean hipotecar(int ip){
        boolean result=false;
        if(encarcelado) return result;
        if(existeLaPropiedad(ip)){
            TituloPropiedad propiedad=propiedades.get(ip);
            result=propiedad.hipotecar(this);
        }
        if(result) Diario.getInstance().ocurreEvento("El jugador "+nombre+" hipoteca "+ip);
        return result;
    }

    private boolean existeLaPropiedad(int ip){
        return ip>=0 && ip<propiedades.size();
    }

    private int getCasasMax(){return CasasMax;}
    private int getHotelesMax(){return HotelesMax;}
    int getCasasPorHotel(){return CasasPorHotel;}

    private boolean puedoEdificarCasa(TituloPropiedad propiedad){
        float precio=propiedad.getPrecioEdificar();
        if(puedoGastar(precio)){
            if(propiedad.getNumCasas()<getCasasMax()){
                return true;
            }
        }
        return false;
    }

    private boolean puedoEdificarHotel(TituloPropiedad propiedad){
        float precio=propiedad.getPrecioEdificar();
        if(puedoGastar(precio)){
            if(propiedad.getNumHoteles()<getHotelesMax()){
                if (propiedad.getNumCasas()>=getCasasPorHotel()){
                    return true;
                }
            }
        }
        return false;
    }

}

File: ruby/lib/jugador.rb
require_relative('titulo_propiedad')
#encoding:utf-8

module Civitas
    class Jugador
        
        @CasasMax=4
        @@CasasPorHotel=4
        @HotelesMax=4
        @@PasoPorSalida=1000
        @@PrecioLibertad=200
        @@SaldoInicial=7500
    
        
        def initialize(nombre)
            @salvoconducto=nil
            @propiedades=[]
            @nombre=nombre
            @numCasillaActual=0
            @puedeComprar=false
            @saldo=@@SaldoInicial
            @encarcelado=false
            #Encarcelado es protected       
        end
        attr_reader :salvoconducto, :propiedades, :nombre, :numCasillaActual, :puedeComprar, :saldo, :encarcelado
        
        #protected
        def copia(otro)
            @salvoconducto=otro.salvoconducto
            @propiedades=otro.propiedades
            @nombre=otro.nombre
            @numCasillaActual=otro.numCasillaActual
            @puedeComprar=otro.puedeComprar
            @encarcelado=otro.encarcelado
            @saldo=otro.saldo
        end

        public
        def cancelarHipoteca(ip)
            result = false

            if (@encarcelado)
                return result
            end

            if existeLaPropiedad(ip)
                propiedad = @propiedades[ip]
                cantidad = propiedad.getImporteCancelarHipoteca
                
                if puedoGastar(cantidad)
                    result = propiedad.cancelarHipoteca(self)

                    if result
                        Diario.instance.ocurreEvento("El jugador #{@nombre} cancela la hipoteca de la propiedad #{ip}")
                    end
                end
            end

            return result
        end

        def cantidadCasasHoteles
            total = 0
            for propiedad in @propiedades
                total += propiedad.cantidadCasasHoteles
            end

            return total
        end

        def comprar(titulo)
            result = false

            if (@encarcelado)
                return result
            end

            if @puedeComprar
                precio = titulo.precioCompra

                if puedoGastar(precio)
                    result = titulo.comprar(self)

                    if (result)
                        @propiedades << titulo
                        Diario.instance.ocurreEvento("El jugador #{@nombre} compra la propiedad #{titulo.toString}")
                    end
                    
                end
                @puedeComprar = false
            end

            return result
        end

        def construirCasa(ip)
            result = false
            if @encarcelado
                return result
            end

            if existeLaPropiedad(ip)
                propiedad = @propiedades[ip]
                puedoEdificar = puedoEdificarCasa(propiedad)

                if puedoEdificar
                    result = propiedad.construirCasa(self)
                    Diario.instance.ocurreEvento("El jugador #{@nombre} construye casa en #{ip}")
                end
            end
        end

        def construirHotel(ip)
            result = false

            if @encarcelado
                return result
            end

            if existeLaPropiedad(ip)
                propiedad = @propiedades[ip]
                puedoEdificar = puedoEdificarHotel(propiedad) #He cambiado el nombre del que hay del diagrama para diferenciarlo de la funcion

                if puedoEdificar
                    result = propiedad.construirHotel(self)
                    propiedad.derruirCasas(@@CasasPorHotel, self)
                    Diario.instance.ocurreEvento("El jugador #{@nombre} construye hotel en su propiedad #{ip}")
                end
            end

            return result
        end

        def enBancarrota
            return @saldo <= 0
        end

        def encarcelar(numCasillaCarcel)
            if debeSerEncarcelado
                moverACasilla(numCasillaCarcel)
                @encarcelado=true
                Diario.instance.ocurreEvento("El jugador "+@nombre+" ha sido encarcelado")
            end
            return @encarcelado
        end

        def getCasasPorHotel
            return @@CasasPorHotel
        end

        def getNumCasillaActual
            return @numCasillaActual
        end

        def getPuedeComprar
            return @puedeComprar
        end

        def moverACasilla(numCasilla)
            if (@encarcelado) 
                return false
            else 
                @numCasillaActual = numCasilla
                @puedeComprar = false
                Diario.instance.ocurreEvento("Se ha desplazado a la casilla: #{numCasilla}")
                return true
            end
        end

        def obtenerSalvoconducto(sorpresa)
            if @encarcelado
                return false            
            else
                @salvoconducto=sorpresa
                Diario.instance.ocurreEvento("El jugador #{@nombre} obtiene salvoconducto")
                return true
            end
        end

        def salirCarcelPagando
            if (@encarcelado && puedeSalirCarcelPagando) 
                paga(@@PrecioLibertad)
                @encarcelado = false
                Diario.instance.ocurreEvento("Jugador: " + @nombre + " ha salido de la carcel PAGANDO")
                return true
            else 
                return false
            end
        end

        def salirCarcelTirando
            if (Dado.instance.salgoDeLaCarcel)
                @encarcelado = false;
                Diario.instance.ocurreEvento("Jugador: " + @nombre + " ha salido de la carcel TIRANDO")
                return true
            else 
                return false
            end
        end

        def tieneAlgoQueGestionar
            return @propiedades.size != 0
        end

        def tieneSalvoConducto
            return @salvoconducto!=nil
        end

        def vender(ip)
            if (!@encarcelado && existeLaPropiedad(ip) && @propiedades[ip].vender(self))
                @propiedades.delete_at(ip)
                Diario.instance.ocurreEvento("La propiedad ha sido vendida")
                return true
            end
            return false    
        end

        private
        def existeLaPropiedad(ip)
            return ip >= 0 && ip < @propiedades.length
        end

        protected
        def self.CasasMax
            return @CasasMax
        end

        def self.HotelesMax
            return @HotelesMax
        end

        private
        def self.PrecioLibertad
            return @@PrecioLibertad
        end

        def self.PasoPorSalida
            return @@PasoPorSalida
        end

        def perderSalvoConducto
            @salvoconducto.usada()
            @salvoconducto=nil
        end

        def puedeSalirCarcelPagando
            return @saldo >= @@PrecioLibertad
        end

        def puedoEdificarCasa(propiedad)
            precio=propiedad.precioEdificar
            if(puedoGastar(precio))
                if(propiedad.numCasas<Jugador.CasasMax)
                    return true
                end
            end
            return false
        end

        def puedoEdificarHotel(propiedad)
            precio=propiedad.precioEdificar
            if(puedoGastar(precio))
                if(propiedad.numHoteles<Jugador.HotelesMax)
                    if (propiedad.numCasas>=@@CasasPorHotel)
                        return true
                    end
                end
            end
            return false
        end

        def puedoGastar(precio)
            if(@encarcelado)
                return false
            end
            return @saldo >= precio
        end

        protected
        def debeSerEncarcelado
            if @encarcelado
                return false
            elsif tieneSalvoConducto
                perderSalvoConducto
                Diario.instance.ocurreEvento("El jugador "+@nombre+" se libra de la cárcel")
                return false
            end

            return true
        end
        
        public
        def compareTo(otro)
            return @saldo <=> otro.saldo
        end

        def hipotecar(ip)
            result = false

            if encarcelado
                return result
            end

            if existeLaPropiedad(ip)
                propiedad = @propiedades[ip]
                result = propiedad.hipotecar(self)
            end

            if result
                Diario.instance.ocurreEvento("El jugador #{@nombre} hipoteca la propiedad #{ip}")
            end
            
            return result
        end

        def modificarSaldo(cantidad)
            @saldo += cantidad
            Diario.instance.ocurreEvento("Se ha modificado el saldo en: " + cantidad.to_s + ", SALDO ACTUAL: " + @saldo.to_s)
            return true
        end

        def paga(cantidad)
            return modificarSaldo(cantidad * -1)
        end

        def pagaAlquiler(cantidad)
            if (@encarcelado) 
                return false 
            end
            return paga(cantidad)
        end

        def pagaImpuesto(cantidad)
            if @encarcelado 
                return false 
            end
            return paga(cantidad)
        end

        def pasaPorSalida
            modificarSaldo(@@PasoPorSalida)
            Diario.instance.ocurreEvento("Jugador: " + @nombre + " ha pasado por la salida")
            return true
        end

        def puedeComprarCasilla
            @puedeComprar = !(@encarcelado)
            return @puedeComprar
        end

        def toString
            string=" Nombre: "+@nombre.to_s
            string+="Salvoconducto: "+@salvoconducto.to_s
            string+=" Cas actual: "+@numCasillaActual.to_s
            string+=" Puede comprar: "+@puedeComprar.to_s
            string+=" Saldo: "+@saldo.to_s
            string+=" Encarcelado: "+@encarcelado.to_s

            substr=""
            for i in @propiedades do
                substr+= (i.nombre+" ")
            end
            string+=" Propiedades: "+substr
        end

        def isEncarcelado
            return @encarcelado
        end

        def recibe(cantidad)
            if (@encarcelado) 
                return false
            else 
                return modificarSaldo(cantidad)
            end
        end

        def getPropiedades_toString
            String res=[]
            for i in @propiedades do
                res << i.nombre
            end
            return res
        end

    end
end